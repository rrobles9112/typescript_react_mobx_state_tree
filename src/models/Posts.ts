import {
  types,
  Instance,
  SnapshotIn,
  getParent,
  destroy,
  flow,
  getSnapshot,
} from "mobx-state-tree";
import { Auth } from "./Auth";
import { rootStore } from "./Root";
export const Post = types.model({
  title: types.string,
  content: types.string,
  image: types.string,
});

export const Posts = types
  .model({
    posts: types.optional(types.array(Post), []),
  })
  .actions((self) => ({
    getPosts: flow(function* fetchPosts() {
      try {
        const {realToken } = getSnapshot<typeof rootStore>(rootStore).auth;
        const response = yield fetch("https://api.joonik.com/posts", {
          method: "GET",
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json",
            Authorization: `Bearer ${realToken}`,
          },
        });
        const result = yield response.json();
        self.posts = result;
        console.log(result);
      } catch (e) {
        console.error("Failed to fetch", e);
      }
    }),
  }));
