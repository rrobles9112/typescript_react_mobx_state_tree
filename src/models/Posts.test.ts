import { Post, Posts } from "./Posts";
import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { reaction } from "mobx";

it("should create a Post instance", () => {
  const item = Post.create({
    title: "Google Pixel 3",
    content: "lorem",
    image: "",
  });
  const patches: any = [];

  onPatch(item, (snapshot) => {
    patches.push(snapshot);
  });

  expect(patches).toMatchSnapshot();
});

it("should create an array Posts instance", () => {
  const PostsInstance = Posts.create({
    posts: [
      {
        title: "Google Pixel 3",
        content: "lorem insump",
        image: "image.png",
      },
      {
        title: "Google Pixel 3",
        content: "lorem insump",
        image: "image.png",
      },
    ],
  });
  expect(getSnapshot(PostsInstance)).toMatchSnapshot();
});

/*

it("can add new items to the cart", () => {
  const cart = Cart.create();
  const states: any = [];

  onSnapshot(cart, (snapshot) => {
    states.push(snapshot);
  });

  cart.addCartItem({
    name: "New Item 1",
    price: 1,
  });
  cart.addCartItem({
    name: "New Item 2",
    price: 2,
  });
  expect(states).toMatchSnapshot();
});

it("can remove an item from the cart", () => {
  const cart = Cart.create();
  const states: any = [];

  onSnapshot(cart, (snapshot) => {
    states.push(snapshot);
  });

  cart.addCartItem({
    name: "New Item 1",
    price: 1,
  });
  cart.addCartItem({
    name: "New Item 2",
    price: 2,
  });
  cart.items[0].remove();
  cart.items[0].remove();
  expect(states).toMatchSnapshot();
});

it("can rename an item in the cart", () => {
  const cart = Cart.create();
  const patches: any = [];

  onPatch(cart, (snapshot) => {
    patches.push(snapshot);
  });

  cart.addCartItem({
    name: "New Item 1",
    price: 1,
  });

  cart.items[0].changeName("New Item (new)");
  expect(patches).toMatchSnapshot();
});

it("can return the total items of the cart", () => {
  const cart = Cart.create();

  cart.addCartItem({
    name: "Item 1",
    price: 1,
  });

  expect(cart.totalItems).toBe(1);
});

it("can calculate the total price of the cart", () => {
  const cart = Cart.create();

  cart.addCartItem({
    name: "Item 1",
    price: 1,
  });
  cart.addCartItem({
    name: "Item 2",
    price: 1,
  });

  expect(cart.totalPrice).toBe(2);
  let changed = 0;
  reaction(
    () => cart.totalPrice,
    () => changed++
  );

  expect(changed).toBe(0);
  cart.items[0].changeName("Item 1 (new)");
  expect(changed).toBe(0);

  cart.items[0].changePrice(2);
  expect(changed).toBe(1);
  expect(cart.totalPrice).toBe(3);
});*/
