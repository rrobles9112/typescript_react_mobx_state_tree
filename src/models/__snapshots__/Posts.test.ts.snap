// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a Post instance 1`] = `Array []`;

exports[`should create an array Posts instance 1`] = `
Object {
  "posts": Array [
    Object {
      "content": "lorem insump",
      "image": "image.png",
      "title": "Google Pixel 3",
    },
    Object {
      "content": "lorem insump",
      "image": "image.png",
      "title": "Google Pixel 3",
    },
  ],
}
`;
